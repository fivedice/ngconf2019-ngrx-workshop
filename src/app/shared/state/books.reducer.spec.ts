import {
  reducer,
  initialState,
  adapter,
  selectAll,
  selectActiveBookId
} from "./books.reducer";
import {
  booksLoaded,
  bookCreated,
  bookDeleted
} from "src/app/books/actions/books-api.actions";

describe("Books Reducer", () => {
  it("should return the initial state when initialized", () => {
    const state = reducer(undefined, { type: "@@init" } as any); // as any due to TS safety
    expect(state).toBe(initialState);
  });

  it("can load all books", () => {
    // Could do this with the adapter too:
    const action = booksLoaded({
      books: [{ id: "1", earnings: 1, name: "something" }]
    });
    const state = reducer(initialState, action);
    expect(state).toMatchSnapshot();
  });

  it("can create a book", () => {
    const action = bookCreated({
      book: { id: "1", earnings: 1, name: "something" }
    });
    const state = reducer(initialState, action);
    expect(state).toMatchSnapshot();
  });

  it("can delete a book", () => {
    const addAction = bookCreated({
      book: { id: "1", earnings: 1, name: "something" }
    });
    const deleteAction = bookDeleted({
      book: { id: "1", earnings: 1, name: "something" }
    });
    const state = [addAction, deleteAction].reduce(reducer, initialState);
    expect(state).toMatchSnapshot();
  });

  describe("Selectors", () => {
    it("can select all books", () => {
      const state = adapter.addAll(
        [{ id: "1", earnings: 1, name: "something" }],
        initialState
      );
      expect(selectAll(state).length).toBe(1);
    });

    it("can return the active book", () => {
      const state = adapter.addAll(
        [{ id: "1", earnings: 1, name: "something" }],
        { ...initialState, activeBookId: "1" }
      );
      expect(selectActiveBookId(state)).toBe("1");
    });
  });
});
