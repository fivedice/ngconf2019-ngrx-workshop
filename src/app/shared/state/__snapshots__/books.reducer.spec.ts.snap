// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books Reducer can create a book 1`] = `
Object {
  "activeBookId": "1",
  "entities": Object {
    "1": Object {
      "earnings": 1,
      "id": "1",
      "name": "something",
    },
  },
  "ids": Array [
    "1",
  ],
}
`;

exports[`Books Reducer can delete a book 1`] = `
Object {
  "activeBookId": null,
  "entities": Object {},
  "ids": Array [],
}
`;

exports[`Books Reducer can load all books 1`] = `
Object {
  "activeBookId": null,
  "entities": Object {
    "1": Object {
      "earnings": 1,
      "id": "1",
      "name": "something",
    },
  },
  "ids": Array [
    "1",
  ],
}
`;
